hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/utils/SparkValidatorUtils.java

```diff
@@ -0,0 +1,169 @@
+      } else {
+        LOG.error("At least one pre-commit validation failed");
+        throw new HoodieValidationException("At least one pre-commit validation failed");
+      }
+    }
+  }
+
+  /**
+   * Run validators in a separate threadpool for parallelism. Each of validator can submit a distributed spark job if needed.
+   */
+  private static CompletableFuture<Boolean> runValidatorAsync(SparkPreCommitValidator validator, HoodieWriteMetadata writeMetadata,
+                                                       Dataset<Row> beforeState, Dataset<Row> afterState, String instantTime) {
+    return CompletableFuture.supplyAsync(() -> {
+      try {
+        validator.validate(instantTime, writeMetadata, beforeState, afterState);
+        LOG.info("validation complete for " + validator.getClass().getName());
+        return true;
+      } catch (HoodieValidationException e) {
+        LOG.error("validation failed for " + validator.getClass().getName());
+        return false;
+      }
+    });
+  }
+
+  /**
+   * Get records from partitions modified as a dataset.
+   * Note that this only works for COW tables.
+   */
+  public static Dataset<Row> getRecordsFromCommittedFiles(SQLContext sqlContext,
+                                                      Set<String> partitionsAffected, HoodieTable table) {
+
+    List<String> committedFiles = partitionsAffected.stream()
+        .flatMap(partition -> table.getBaseFileOnlyView().getLatestBaseFiles(partition).map(bf -> bf.getPath()))
+        .collect(Collectors.toList());
+
+    if (committedFiles.isEmpty()) {
+      return sqlContext.emptyDataFrame();
+    }
+    return readRecordsForBaseFiles(sqlContext, committedFiles);
+  }
+
+  /**
+   * Get records from specified list of data files.
+   */
+  public static Dataset<Row> readRecordsForBaseFiles(SQLContext sqlContext, List<String> baseFilePaths) {
+    return sqlContext.read().parquet(JavaConverters.asScalaBufferConverter(baseFilePaths).asScala());
+  }
+
+  /**
+   * Get reads from paritions modified including any inflight commits.
+   * Note that this only works for COW tables
+   */
+  public static Dataset<Row> getRecordsFromPendingCommits(SQLContext sqlContext, 
+                                                          Set<String> partitionsAffected, 
+                                                          Map<String, List<String>> partitionToReplacedFileIds, 
+                                                          HoodieTable table) {
+
+    // build file system view with pending commits
+    HoodieTableFileSystemView fsView = new HoodieTableFileSystemView(table.getMetaClient(),
+        table.getActiveTimeline().getCommitsTimeline());
+
+    List<String> newFiles = partitionsAffected.stream().flatMap(partition -> {
+      return fsView.getLatestBaseFiles(partition)
+          .filter(bf -> !partitionToReplacedFileIds.getOrDefault(partition, Collections.emptyList()).contains(bf.getFileId()))
```

Comment:
This partitionToReplacedFileIds is not visible to timeline/FileSystemView because it is inflight operation. Its only available in HoodieWriteMetadata returned by write operation.

Reply:
Right. What I am proposing here is to have a class that inherits HoodieTableFileSystemView (something like HoodieTablePreCommitFileSystemView) which takes in as argument existing HoodieTableFileSystemView (committed data) and also uncommitted metadata (HoodieWriteMetadata).  This way of composing will also work even if multiple writes (inflights) are happening concurrently as HoodieTablePreCommitFileSystemView would automatically skip the inflight it sees from other parallel write.

Reply:
HoodieWriteMetadata is in hudi-client. All filesystem views are in hudi-common. Is it fine to add this new HoodieTablePreCommitFileSystemView in hudi-client?

minor: I also feel this abstraction is little complex because this is only used for precommit validations. FileSystemViews also expose lot of methods which (I think) can be misused for HoodieTablePreCommitFileSystemView.

Reply:
@satishkotha : You can pass HoodieCommitMetadata and partitionToReplaceFileIds directly as constructor arguments to HoodieTablePreCommitFileSystemView. As the current setup wont work with concurrent writers, this change is needed for correctness.  Implementing at the filesystemView abstraction is safe and clean as it would hide the callers from needing to worry about inflights. Also, we can have this view be reused.

Reply:
ok. modified. PTAL